### PR TITLE
Updated TwabController and made the interface safer

### PR DESCRIPTION
**Pull request by asselstine**

### CODE REVIEW

File: src/libraries/TwabLib.sol [hunk 1]
```diff
@@ -252,72 +256,92 @@ library TwabLib {
...
   function getTwabBetween(
+    uint32 PERIOD_LENGTH,
     uint32 PERIOD_OFFSET,
     ObservationLib.Observation[MAX_CARDINALITY] storage _observations,
     AccountDetails memory _accountDetails,
     uint32 _startTime,
     uint32 _endTime
   ) internal view returns (uint256) {
+    // The current period can still be changed; so the start of the period marks the beginning of unsafe timestamps.
+    uint32 overwritePeriodStartTime = _overwritePeriodStartTime(PERIOD_LENGTH, PERIOD_OFFSET);
```
Comment (asselstine) [inline src/libraries/TwabLib.sol:307]: Whoops… realized this line should be passing the end time to compute the overwrite start time.  It shouldn’t be using the current time

File: src/TwabController.sol [hunk 2]
```diff
@@ -362,86 +364,12 @@ contract TwabController {
...
-    return
-      TwabLib.isTimeRangeSafe(
-        PERIOD_LENGTH,
-        PERIOD_OFFSET,
-        account.observations,
-        account.details,
-        startTime,
-        endTime
-      );
+  function isDuringOverwritePeriod(uint32 time) external view returns (bool) {
```
Comment (asselstine) [inline src/TwabController.sol]: This name is wrong; it's actually doing the opposite

File: src/libraries/TwabLib.sol [hunk 3]
```diff
@@ -252,72 +256,92 @@ library TwabLib {
...
+
+    // The start time must be older than the current period start time, so that it cannot be changed.
+    uint32 safeStartTime = _startTime < overwritePeriodStartTime ? _startTime : overwritePeriodStartTime - 1;
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Would be useful to create an helper function to retrieve the safe time.

File: src/libraries/TwabLib.sol [hunk 4]
```diff
@@ -252,72 +256,92 @@ library TwabLib {
...
       PERIOD_OFFSET,
       _observations,
       _accountDetails,
-      _endTime
+      safeEndTime
     );
 
-    if (startObservation.timestamp != _startTime) {
-      startObservation = _calculateTemporaryObservation(startObservation, _startTime);
+    if (startObservation.timestamp != safeEndTime) {
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Should be `safeStartTime` no?

File: src/libraries/TwabLib.sol [hunk 5]
```diff
@@ -390,6 +417,20 @@ library TwabLib {
...
+   * @notice Computes the start time of the current period
+   * @param PERIOD_LENGTH The length of an overwrite period
+   * @param PERIOD_OFFSET The offset of the first period
+   */
+  function _overwritePeriodStartTime(
+    uint32 PERIOD_LENGTH,
+    uint32 PERIOD_OFFSET
+  ) private view returns (uint32) {
+    uint32 currentTime = uint32(block.timestamp);
+    uint32 currentPeriod = getTimestampPeriod(PERIOD_LENGTH, PERIOD_OFFSET, currentTime);
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Can be inlined.

File: src/libraries/TwabLib.sol [hunk 6]
```diff
@@ -390,6 +417,20 @@ library TwabLib {
...
+  /**
+   * @notice Computes the start time of the current period
+   * @param PERIOD_LENGTH The length of an overwrite period
+   * @param PERIOD_OFFSET The offset of the first period
+   */
+  function _overwritePeriodStartTime(
+    uint32 PERIOD_LENGTH,
+    uint32 PERIOD_OFFSET
+  ) private view returns (uint32) {
+    uint32 currentTime = uint32(block.timestamp);
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Should use `endTime` instead of `currentTime`.

File: src/libraries/TwabLib.sol [hunk 7]
```diff
@@ -437,17 +491,37 @@ library TwabLib {
     uint32 PERIOD_OFFSET,
     uint32 _timestamp
   ) private pure returns (uint32 period) {
-    if (_timestamp <= PERIOD_OFFSET) {
+    if (_timestamp < PERIOD_OFFSET) {
       return 0;
     }
     // Shrink by 1 to ensure periods end on a multiple of PERIOD_LENGTH.
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Remove this comment since we removed the `-1`.

File: src/libraries/TwabLib.sol [hunk 8]
```diff
@@ -437,17 +491,37 @@ library TwabLib {
...
     // Increase by 1 to start periods at # 1.
-    return ((_timestamp - PERIOD_OFFSET - 1) / PERIOD_LENGTH) + 1;
+    return (_timestamp - PERIOD_OFFSET) / PERIOD_LENGTH + 1;
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Why was `- 1` removed?

File: src/libraries/TwabLib.sol [hunk 9]
```diff
@@ -493,18 +567,14 @@ library TwabLib {
       return prevOrAtObservation;
     }
 
-    // If there is only 1 actual observation, either return that observation or a zeroed observation
+    // If there is only 1 observation and it's after the target, then return zero
     if (_accountDetails.cardinality == 1) {
-      if (_targetTime >= prevOrAtObservation.timestamp) {
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol:498]: We removed this condition cause it's supposed to be handled in the condition above right?

File: src/libraries/TwabLib.sol [hunk 10]
```diff
@@ -533,223 +603,35 @@ library TwabLib {
...
-   * @param _observations The circular buffer of observations
-   * @param _accountDetails The account details to query with
+   * @dev A timestamp is safe if it is before the current overwrite period
+   * @param PERIOD_LENGTH The period length to use to calculate the period
+   * @param PERIOD_OFFSET The period offset to use to calculate the period
    * @param _time The timestamp to check
    * @return isSafe Whether or not the timestamp is safe
    */
-  function isTimeSafe(
+  function isDuringOverwritePeriod(
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol]: Rename to `isBeforeOverwritePeriod`.

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: src/TwabController.sol [hunk 11]
```diff
@@ -232,7 +232,7 @@ contract TwabController {
     uint32 targetTime
   ) external view returns (uint256) {
     TwabLib.Account storage _account = userObservations[vault][user];
-    return TwabLib.getBalanceAt(PERIOD_OFFSET, _account.observations, _account.details, targetTime);
+    return TwabLib.getBalanceAt(PERIOD_LENGTH, PERIOD_OFFSET, _account.observations, _account.details, _periodEndOnOrAfter(targetTime));
```
Comment (PierrickGT) [inline src/TwabController.sol]: So in `getBalanceAt`, `targetTime` is now always the end of the targeted period?
Maybe the `targetTime` param in TwabLib should be renamed `targetPeriodEndTime`.

File: src/TwabController.sol [hunk 12]
```diff
@@ -262,13 +262,16 @@ contract TwabController {
...
     return
       TwabLib.getTwabBetween(
+        PERIOD_LENGTH,
         PERIOD_OFFSET,
         _account.observations,
         _account.details,
-        startTime,
-        endTime
+        _periodEndOnOrAfter(startTime),
+        _periodEndOnOrAfter(endTime)
```
Comment (PierrickGT) [inline src/TwabController.sol:274]: Same here, we could rename the last two params in `TwabLib.getTwabBetween`.

File: src/TwabController.sol [hunk 13]
```diff
@@ -286,16 +289,44 @@ contract TwabController {
...
       );
   }
 
+  /**
+   * @notice Computes the period end timestamp on or after the given timestamp.
+   * @param _timestamp The timestamp to compute the period start time for
+   * @return The start time of the period in which the given timestamp falls
```
Comment (PierrickGT) [inline src/TwabController.sol]:
```suggestion
   * @param _timestamp The timestamp to compute the period end time for
   * @return The end time of the period in which the given timestamp falls
```

File: src/TwabController.sol [hunk 14]
```diff
@@ -286,16 +289,44 @@ contract TwabController {
...
+   * @return A period end time.
+   */
+  function _periodEndOnOrAfter(uint32 _timestamp) internal view returns (uint32) {
+    if (_timestamp < PERIOD_OFFSET) {
+      return PERIOD_OFFSET;
+    }
+    if ((_timestamp - PERIOD_OFFSET) % PERIOD_LENGTH == 0) {
+      return _timestamp;
+    }
+    uint32 period = TwabLib.getTimestampPeriod(PERIOD_LENGTH, PERIOD_OFFSET, _timestamp);
```
Comment (PierrickGT) [inline src/TwabController.sol]: Can be inlined.

File: src/TwabController.sol [hunk 15]
```diff
@@ -362,86 +393,21 @@ contract TwabController {
...
   /**
-   * @notice Checks if the given time range is safe to perform a historic balance lookup on.
-   * @dev A timestamp is safe if it is between (or at) the newest observation in a period and the end of the period.
-   * @dev If the endtime being queried is in a period that has not yet ended, the output for this function may change.
-   * @param vault The vault to check
-   * @param startTime The start of the timerange to check
-   * @param endTime The end of the timerange to check
-   * @return isSafe Whether or not the time range is safe
+   * @notice Computes the timestamp at which the current overwrite period started.
+   * The overwrite period is the period during which observations are collated.
```
Comment (PierrickGT) [inline src/TwabController.sol]:
```suggestion
   * @dev The overwrite period is the period during which observations are collated.
```

File: src/libraries/TwabLib.sol [hunk 16]
```diff
@@ -390,6 +414,20 @@ library TwabLib {
...
+   * @notice Computes the start time of the current overwrite period
+   * @param PERIOD_LENGTH The length of an overwrite period
+   * @param PERIOD_OFFSET The offset of the first period
+   * @return The start time of the current overwrite period
+   */
+  function _currentOverwritePeriodStartedAt(
+    uint32 PERIOD_LENGTH,
+    uint32 PERIOD_OFFSET
+  ) private view returns (uint32) {
+    uint32 period = getTimestampPeriod(PERIOD_LENGTH, PERIOD_OFFSET, uint32(block.timestamp));
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol:427]: Can be inlined.

File: src/libraries/TwabLib.sol [hunk 17]
```diff
@@ -399,19 +437,19 @@ library TwabLib {
...
     return
       _observation.cumulativeBalance +
       uint128(_observation.balance) *
-      (_timestamp.checkedSub(_observation.timestamp, _timestamp));
+      (_timestamp.checkedSub(_observation.timestamp, uint32(block.timestamp)));
   }
 
   /**
    * @notice Calculates the period a timestamp falls within.
    * @dev All timestamps prior to the PERIOD_OFFSET fall within period 0.
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol:450]: They fall within period 0 cause before `PERIOD_OFFSET`, a user potentially has a recorded balance?

File: src/libraries/TwabLib.sol [hunk 18]
```diff
@@ -533,223 +608,52 @@ library TwabLib {
...
-  ) internal view returns (bool) {
-    return
-      _isTimeSafe(PERIOD_LENGTH, PERIOD_OFFSET, _observations, _accountDetails, _startTime) &&
-      _isTimeSafe(PERIOD_LENGTH, PERIOD_OFFSET, _observations, _accountDetails, _endTime);
+  modifier requireFinalized(uint32 PERIOD_LENGTH, uint32 PERIOD_OFFSET, uint256 _timestamp) {
+    // The current period can still be changed; so the start of the period marks the beginning of unsafe timestamps.
+    uint32 overwritePeriodStartTime = _currentOverwritePeriodStartedAt(PERIOD_LENGTH, PERIOD_OFFSET);
+    // timestamp == overwritePeriodStartTime doesn't matter, because the cumulative balance won't be impacted
+    if (_timestamp > overwritePeriodStartTime) {
+      revert TimestampNotFinalized(_timestamp, overwritePeriodStartTime);
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol:655]: Would make more sense if the error was `PeriodNotFinalized`.
```suggestion
      revert PeriodNotFinalized(_timestamp, overwritePeriodStartTime);
```

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: src/TwabController.sol (hunk 11 above)
Comment (asselstine) [reply src/TwabController.sol]: Yes, targetTime is always "snapped" to the period that ends on it or ends after it.

I feel like `targetPeriodEndTime` isn't quite specific enough.  Perhaps `periodEndOnOrAfterTime`

File: src/TwabController.sol (hunk 12 above)
Comment (asselstine) [reply src/TwabController.sol:274]: I'll add additional comments for these ones

**Pull request merged**